particle-iot/particle-cli
# Feature/asset bundle

## Description

Adds support to bundle assets in the CLI using the `particle bundle` command
### Usage
`particle bundle application.bin --assets /path/to/assets/folder --saveTo myBundle.zip`
- Providing `application.bin` is mandatory
- If `--assets` option is not specified, the files from the folder named `assets` in the current directory are picked up
    - :warning: A caveat with testing in particle-cli root directory is that there is an existing `assets` directory irrelevant to this implementation, and that might get picked up
- If `--saveTo` is not specified, default name is used

## How to Test

### Setup
1. Pull down the branch: `git pull && git checkout feature/asset-bundle`
2. Run `nvm use`
3. Reinstall dependencies: `npm run reinstall`
4. Run tests: `npm test`

### Creating a bundle
5. Save files of your choice to bundle in a folder (example: `/path/to/temp_assets/`
6. Specify the `application.bin` and the folder using the command 
```
particle-cli (feature/asset-bundle) $ npm start -- bundle /path/to/application.bin --assets /path/to/temp_assets/ --saveTo myBundle.zip
```
7. Verify that the bundle is created in the same directory
8. Verify from the CLI output that the expected files are included in the bundle

### Testing scenarios
1. Add special files like `__MACOS` or `.xyz` into the assets folder and verify that they are _not_ included in the bundle
2. Add nested directories in side the /assets folder and verify that they are _not_ included in the bundle
3. Please thoroughly test any additional customer-driven edge cases

## Related Issues / Discussions

[117747](https://app.shortcut.com/particle/story/117747/create-zipped-app-bundle)

## Completeness

- [] CI is passing
